openstack/identity/v3/registered_limits/requests.go

```diff
@@ -0,0 +1,159 @@
+	})
+}
+
+// BatchCreateOptsBuilder allows extensions to add additional parameters to
+// the Create request.
+type BatchCreateOptsBuilder interface {
+	ToRegisteredLimitsCreateMap() (map[string]interface{}, error)
+}
+
+type CreateOpts struct {
+	// RegionID is the ID of the region where the limit is applied.
+	RegionID string `json:"region_id,omitempty"`
+
+	// ServiceID is the ID of the service where the limit is applied.
+	ServiceID string `json:"service_id" required:"true"`
+
+	// Description of the limit.
+	Description string `json:"description,omitempty"`
+
+	// ResourceName is the name of the resource that the limit is applied to.
+	ResourceName string `json:"resource_name" required:"true"`
+
+	// DefaultLimit is the default limit.
+	DefaultLimit int `json:"default_limit"`
+}
+
+// BatchCreateOpts provides options used to create limits.
+type BatchCreateOpts []CreateOpts
+
+// ToRegisteredLimitsCreateMap formats a BatchCreateOpts into a create request.
+func (opts BatchCreateOpts) ToRegisteredLimitsCreateMap() (map[string]interface{}, error) {
+	registered_limits := make([]map[string]interface{}, len(opts))
+	for i, registered_limit := range opts {
+		registeredLimitMap, err := registered_limit.ToMap()
+		if err != nil {
+			return nil, err
+		}
+		registered_limits[i] = registeredLimitMap
+	}
+	return map[string]interface{}{"registered_limits": registered_limits}, nil
+}
+
+func (opts CreateOpts) ToMap() (map[string]interface{}, error) {
+	return gophercloud.BuildRequestBody(opts, "")
+}
+
+// BatchCreate creates new Limits.
+func BatchCreate(client *gophercloud.ServiceClient, opts BatchCreateOptsBuilder) (r CreateResult) {
+	b, err := opts.ToRegisteredLimitsCreateMap()
+	if err != nil {
+		r.Err = err
+		return
+	}
+	resp, err := client.Post(rootURL(client), &b, &r.Body, &gophercloud.RequestOpts{
+		OkCodes: []int{201},
+	})
+	_, r.Header, r.Err = gophercloud.ParseResponse(resp, err)
+	return
+}
+
+// Get retrieves details on a single registered_limit, by ID.
+func Get(client *gophercloud.ServiceClient, registeredLimitID string) (r GetResult) {
+	resp, err := client.Get(resourceURL(client, registeredLimitID), &r.Body, nil)
+	_, r.Header, r.Err = gophercloud.ParseResponse(resp, err)
+	return
+}
+
+// UpdateOptsBuilder allows extensions to add additional parameters to
+// the Update request.
+type UpdateOptsBuilder interface {
+	ToRegisteredLimitUpdateMap() (map[string]interface{}, error)
+}
+
+// UpdateOpts represents parameters to update a domain.
+type UpdateOpts struct {
+	// Description of the registered_limit.
+	Description *string `json:"description,omitempty"`
+
+	// DefaultLimit is the override limit.
+	DefaultLimit *int `json:"default_limit,omitempty"`
+
+	// RegionID is the ID of the region where the limit is applied.
+	RegionID string `json:"region_id,omitempty"`
+
+	// ServiceID is the ID of the service where the limit is applied.
+	ServiceID string `json:"service_id" required:"true"`
```

Comment:
this shouldnt be `required` based on docs.

I do read this on docs though:
> Either service_id, resource_name, or region_id must be different than existing value otherwise it will raise 409.

And i have 0 clue why they have added this restriction, nor why they allow updates on ServiceID or ResourceName... :man_shrugging:

Reply:
Same, I don't have clue as well. I think this is global default limit across all the projects and if you want different limit for particular service then you can override it by "limit" api. (This is my guess though)

Reply:
Local acceptance test results;


=== RUN   TestRegisteredLimitsList
--- PASS: TestRegisteredLimitsList (0.14s)
=== RUN   TestCreateRegisteredLimits
--- PASS: TestCreateRegisteredLimits (0.33s)
=== RUN   TestRegisteredLimitsCRUD
    tools.go:80: {
          "id": "f71aca4f1d82428488ee230a9ee123a4",
          "region_id": "RegionOne",
          "service_id": "4da42dd1adf54bbb878d925fc64382da",
          "description": "TESTLIMITS-DESC-Zs7SnB5c",
          "resource_name": "LIMIT-NAME-t22jpMnh",
          "default_limit": 55,
          "links": {
            "self": "http://10.0.108.117/identity/v3/registered_limits/f71aca4f1d82428488ee230a9ee123a4"
          }
        }
    tools.go:80: {
          "id": "f71aca4f1d82428488ee230a9ee123a4",
          "region_id": "RegionOne",
          "service_id": "4da42dd1adf54bbb878d925fc64382da",
          "description": "TESTLIMITS-DESC-Zs7SnB5c",
          "resource_name": "LIMIT-NAME-t22jpMnh",
          "default_limit": 55,
          "links": {
            "self": "http://10.0.108.117/identity/v3/registered_limits/f71aca4f1d82428488ee230a9ee123a4"
          }
        }
    tools.go:80: {
          "id": "f71aca4f1d82428488ee230a9ee123a4",
          "region_id": "RegionOne",
          "service_id": "4da42dd1adf54bbb878d925fc64382da",
          "description": "Test description for registered limit",
          "resource_name": "LIMIT-NAME-MyZR9stf",
          "default_limit": 1000,
          "links": {
            "self": "http://10.0.108.117/identity/v3/registered_limits/f71aca4f1d82428488ee230a9ee123a4"
          }
        }
--- PASS: TestRegisteredLimitsCRUD (0.59s)



Local Unit test results;
go test -v ./openstack/identity/v3/registeredlimits/testing
=== RUN   TestListRegisteredLimits
--- PASS: TestListRegisteredLimits (0.00s)
=== RUN   TestListRegisteredLimitsAllPages
--- PASS: TestListRegisteredLimitsAllPages (0.00s)
=== RUN   TestCreateRegisteredLimits
--- PASS: TestCreateRegisteredLimits (0.00s)
=== RUN   TestGetRegisteredLimit
--- PASS: TestGetRegisteredLimit (0.00s)
=== RUN   TestDeleteRegisteredLimit
--- PASS: TestDeleteRegisteredLimit (0.00s)
=== RUN   TestUpdateRegisteredLimit
--- PASS: TestUpdateRegisteredLimit (0.00s)
PASS
ok  	github.com/gophercloud/gophercloud/openstack/identity/v3/registeredlimits/testing	0.009s